### PR TITLE
Threaded Sensitivity Analysis

### PR DESCRIPTION
**Pull request by mjohnson541**
This pull request implements multithreaded sensitivity analysis.

### CODE REVIEW

File: src/ThreadedSensitivities.jl [hunk 1]
```diff
@@ -1,3 +1,36 @@
...
+    y0 = react.y0
+    tspan = react.tspan
+    p = react.p
+    for i in 1:length(p)
+        r = deepcopy(react)
+        jacy!(J::Q2,y::T,p::V,t::Q) where {Q2,T,Q<:Real,V} = jacobiany!(J,y,p,t,r.domain,r.interfaces,nothing)
+        jacp!(J::Q2,y::T,p::V,t::Q) where {Q2,T,Q<:Real,V} = jacobianp!(J,y,p,t,r.domain,r.interfaces,nothing)
+
+        function dsdt(p_index)
+                function dsdt!(ds, s, local_params, t)
```
Comment (hwpang) [inline src/ThreadedSensitivities.jl:99]: Can you fix the indentation here? Why do we wrap `dsdt!` in `dsdt`?

File: src/ThreadedSensitivities.jl [hunk 2]
```diff
@@ -1,3 +1,36 @@
...
+                jy = zeros(length(y0), length(y0))
+                jp = zeros(length(y0), length(p))
+                jacy!(jy, sol(t), p, t)
+                jacp!(jp, sol(t), p, t)
+                c = jp[:, p_index]
```
Comment (hwpang) [inline src/ThreadedSensitivities.jl:104]: Can you add @views and @inbounds

File: src/ThreadedSensitivities.jl [hunk 3]
```diff
@@ -1,3 +1,36 @@
...
+        r = deepcopy(react)
+        jacy!(J::Q2,y::T,p::V,t::Q) where {Q2,T,Q<:Real,V} = jacobiany!(J,y,p,t,r.domain,r.interfaces,nothing)
+        jacp!(J::Q2,y::T,p::V,t::Q) where {Q2,T,Q<:Real,V} = jacobianp!(J,y,p,t,r.domain,r.interfaces,nothing)
+
+        function dsdt(p_index)
+                function dsdt!(ds, s, local_params, t)
+                jy = zeros(length(y0), length(y0))
+                jp = zeros(length(y0), length(p))
+                jacy!(jy, sol(t), p, t)
+                jacp!(jp, sol(t), p, t)
```
Comment (hwpang) [inline src/ThreadedSensitivities.jl:103]: Don't sure how much it matters but can you save `sol(t)` and avoid calling it multiple times here?

File: src/ThreadedSensitivities.jl [hunk 4]
```diff
@@ -1,3 +1,36 @@
...
+"""
+generate individual sensitivity ODEs for each parameter
+"""
+function generatesensitivityodes(react,sol)
+    sa_list = []
+    y0 = react.y0
+    tspan = react.tspan
+    p = react.p
+    for i in 1:length(p)
+        r = deepcopy(react)
```
Comment (hwpang) [inline src/ThreadedSensitivities.jl:94]: Why do we need to deepcopy this? Is the reactor changed somehow?

File: src/ThreadedSensitivities.jl [hunk 5]
```diff
@@ -34,3 +34,20 @@ function generatesensitivityodes(react,sol)
...
+"""
+Combine ODE solutions into a sensitivity solution
+"""
+function generatesenssolution(sol,sensdict,sensprob)
+    ts = sol.t
+    ordkeys = sort([x for x in keys(sensdict)])
+    u = deepcopy(sol.u)
+    for k in ordkeys
+        for i in 1:length(u)
+        u[i] = vcat(u[i],sensdict[k](ts[i]))
```
Comment (hwpang) [inline src/ThreadedSensitivities.jl:128]: Can you fix the indentation

File: src/ThreadedSensitivities.jl [hunk 6]
```diff
@@ -34,3 +34,20 @@ function generatesensitivityodes(react,sol)
     end
     return sa_list
 end
+
+"""
+Combine ODE solutions into a sensitivity solution
+"""
+function generatesenssolution(sol,sensdict,sensprob)
+    ts = sol.t
+    ordkeys = sort([x for x in keys(sensdict)])
```
Comment (hwpang) [inline src/ThreadedSensitivities.jl:124]: I feel the list comprehension is not needed, can we use `sort(list(keys(sensdict)))` or something like this?

File: src/ThreadedSensitivities.jl (hunk 4 above)
Comment (mjohnson541) [reply src/ThreadedSensitivities.jl:94]: This deepcopy is the fix for the bug that took Mark and I so long to find. The issue is that some of our domains precompute kfs and krevs, but in gas phase these arrays need modified to account for efficiencies. So if you don't copy  it you have multiple threads trying to modify the same array. It's definitely possible to reduce the copy time by being more clever and careful, but there's some caveats particularly related to our use of SparseArrays that make it non-trivial to do that.

File: src/ThreadedSensitivities.jl (hunk 2 above)
Comment (mjohnson541) [reply src/ThreadedSensitivities.jl:104]: Yep

File: src/ThreadedSensitivities.jl (hunk 3 above)
Comment (mjohnson541) [reply src/ThreadedSensitivities.jl:103]: Yep

File: src/ThreadedSensitivities.jl (hunk 5 above)
Comment (mjohnson541) [reply src/ThreadedSensitivities.jl:128]: Yep

File: src/ThreadedSensitivities.jl (hunk 6 above)
Comment (mjohnson541) [reply src/ThreadedSensitivities.jl:124]: `Array` doesn't work at least. I'm not sure if there's a better more readable way than that.

File: src/ThreadedSensitivities.jl (hunk 1 above)
Comment (mjohnson541) [reply src/ThreadedSensitivities.jl]: Yep! That was how Mark had it setup, but I've unwrapped it now.

**Comment by mjohnson541** [comment]
Okay, I think I've addressed all of the comments and tests have passed.

**Review by hwpang — APPROVED** [review]

**Pull request merged**